frmts/sentinel2/sentinel2dataset.cpp

```diff
@@ -438,6 +441,27 @@ int SENTINEL2Dataset::Identify( GDALOpenInfo *poOpenInfo )
         strstr(pszHeader, "User_Product_Level-2A" ) != nullptr )
         return TRUE;
 
+    /* Last effort – look inside of a zip file for metadata file with
+     * matching name. */
+    if ( EQUAL(CPLGetExtension(pszJustFilename), "zip") )
+    {
+        CPLString osFilename(poOpenInfo->pszFilename);
+        if( strncmp(osFilename, "/vsizip/", strlen("/vsizip/")) != 0 )
+            osFilename = "/vsizip/" + osFilename;
+
+        auto psDir = VSIOpenDir(osFilename.c_str(), 1, nullptr);
+        while ( const VSIDIREntry* psEntry = VSIGetNextDirEntry(psDir) )
```

Comment:
This causes a crash in autotest/gdrivers/jpeg.py::test_jpeg_27_max_memory  , that tries to open a '/vsisubfile/146,/vsizip/../gcore/data/eofloop_valid_huff.tif.zip' file .
The issue is that putting a '/vsizip/' prefix in front of that filename results in something that can't be opened by VSIOpenDir().  psDir should thus be tested against nullptr before being used. Applies here and at line 629 as well.

Reply:
Unfortunately on Debian Sid I'm unable to run full test suite :-(
I reshaped my code according to your suggestions – should be better now.